x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -237,4 +392,293 @@ public void onFailure(Exception e) {
+     */
+    private class PeriodicMaintenanceTask implements Runnable, Releasable {
+
+        private final TimeValue keepAlive;
+        private final int batchSize;
+
+        private final AtomicReference<Exception> error = new AtomicReference<>();
+        private final AtomicBoolean closed = new AtomicBoolean();
+        private final AtomicLong deletes = new AtomicLong();
+        private final AtomicLong total = new AtomicLong();
+
+        private volatile Map<String, Set<String>> existingSnapshots;
+        private volatile SearchResponse searchResponse;
+        private volatile String pointIntTimeId;
+        private volatile Object[] searchAfter;
+
+        PeriodicMaintenanceTask(TimeValue keepAlive, int batchSize) {
+            this.keepAlive = keepAlive;
+            this.batchSize = batchSize;
+        }
+
+        @Override
+        public void run() {
+            final PeriodicMaintenanceTask maintenanceTask = this;
+            assert assertGenericThread();
+
+            try {
+                ensureOpen();
+                if (pointIntTimeId == null) {
+                    final OpenPointInTimeRequest openRequest = new OpenPointInTimeRequest(SNAPSHOT_BLOB_CACHE_INDEX);
+                    openRequest.keepAlive(keepAlive);
+                    clientWithOrigin.execute(OpenPointInTimeAction.INSTANCE, openRequest, new ActionListener<>() {
+                        @Override
+                        public void onResponse(OpenPointInTimeResponse response) {
+                            logger.trace("periodic maintenance task initialized with point-in-time id [{}]", response.getPointInTimeId());
+                            maintenanceTask.pointIntTimeId = response.getPointInTimeId();
+                            executeNext(maintenanceTask);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            if (TransportActions.isShardNotAvailableException(e)) {
+                                complete(maintenanceTask, null);
+                            } else {
+                                complete(maintenanceTask, e);
+                            }
+                        }
+                    });
+                    return;
+                }
+
+                final String pitId = pointIntTimeId;
+                assert Strings.hasLength(pitId);
+
+                if (searchResponse == null) {
+                    final SearchSourceBuilder searchSource = new SearchSourceBuilder();
+                    searchSource.trackScores(false);
+                    searchSource.sort("_shard_doc");
+                    searchSource.size(batchSize);
+                    final PointInTimeBuilder pointInTime = new PointInTimeBuilder(pitId);
+                    searchSource.pointInTimeBuilder(pointInTime);
+                    pointInTime.setKeepAlive(keepAlive);
```

Comment:
Perhaps swap the order of these two lines?

Reply:
Sure, pushed in 2a84b4a65854a586054cbe73fad7192e5815a0ef